### PR TITLE
PLUGIN-1609: Kafka Batch Sink | Add null check for key when key is missing or null in the message when using message key field

### PR DESCRIPTION
**Pull request by sgarg-CS**
[PLUGIN-1609](https://cdap.atlassian.net/browse/PLUGIN-1609)

[PLUGIN-1609]: https://cdap.atlassian.net/browse/PLUGIN-1609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java [hunk 1]
```diff
@@ -118,7 +118,13 @@ public void transform(StructuredRecord input, Emitter<KeyValue<Text, Text>> emit
         emitter.emit(new KeyValue<>((Text) null, new Text(body)));
       } else {
         String key = input.get(producerConfig.key);
-        emitter.emit(new KeyValue<>(new Text(key), new Text(body)));
+        // The key inside the message may be null if it is not specified while publishing the event.
+        // https://stackoverflow.com/questions/29511521
+        if (key != null) {
+          emitter.emit(new KeyValue<>(new Text(key), new Text(body)));
+        } else {
+          emitter.emit(new KeyValue<>((Text) null, new Text(body)));
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: It may be better to fail here with a clear exception message. If the null was not intended, this will affect the distribution of data.

Comment (sgarg-CS) [reply kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: Throwing Runtime Exception with the reason for failure 4ab91e8c093be7719ddf68b444177b0c0e029c87

File: kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java [hunk 2]
```diff
@@ -118,7 +118,14 @@ public void transform(StructuredRecord input, Emitter<KeyValue<Text, Text>> emit
         emitter.emit(new KeyValue<>((Text) null, new Text(body)));
       } else {
         String key = input.get(producerConfig.key);
-        emitter.emit(new KeyValue<>(new Text(key), new Text(body)));
+        // The key inside the message may be null if it is not specified while publishing the event.
+        // https://stackoverflow.com/questions/29511521
+        if (key != null) {
+          emitter.emit(new KeyValue<>(new Text(key), new Text(body)));
+        } else {
+          throw new RuntimeException(String.format("Key for the event: %s published into Kafka must exist in the " +
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: nit: Change the message to indicate that the key field value is null.

Comment (greeshmaswaminathan) [reply kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: Can a unit test be added for this ?

Comment (sgarg-CS) [reply kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: > nit: Change the message to indicate that the key field value is null.


Updated the message. 748e8da5d85340e3d6315e016af4a347e354a150

Comment (sgarg-CS) [reply kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: > Can a unit test be added for this ?

I'm working on adding a unit test for the `null` key use case

Comment (sgarg-CS) [reply kafka-plugins-client/src/main/java/io/cdap/plugin/kafka/sink/KafkaBatchSink.java]: Unit test added 399af3d6b9ad9fc47f2234d2790018cfd584144f

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java [hunk 3]
```diff
@@ -228,7 +229,94 @@ public Alert apply(String s) {
...
+      .addStage(alert)
+      .addConnection(source.getName(), transform.getName())
+      .addConnection(transform.getName(), sink.getName())
+      .addConnection(transform.getName(), alert.getName())
+      .build();
+
+    // create the pipeline
+    ApplicationId pipelineId = NamespaceId.DEFAULT.app("testKafkaSink");
+    ApplicationManager appManager = deployApplication(pipelineId, new AppRequest<>(APP_ARTIFACT, pipelineConfig));
+
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: nit: Remove extra newlines.

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java [hunk 4]
```diff
@@ -228,7 +229,94 @@ public Alert apply(String s) {
...
+    sinkProperties.put("compressionType", "none");
+
+    Map<String, String> alertProperties = new HashMap<>();
+    alertProperties.put("brokers", "localhost:" + kafkaPort);
+    alertProperties.put("topic", alertsTopic);
+
+    ETLStage source = new ETLStage("source", MockSource.getPlugin(inputName));
+    ETLStage sink =
+      new ETLStage("sink", new ETLPlugin("Kafka", KafkaBatchSink.PLUGIN_TYPE, sinkProperties, null));
+    ETLStage transform = new ETLStage("nullAlert", NullAlertTransform.getPlugin("id"));
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Any reason for  using an alert transform here ?

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java [hunk 5]
```diff
@@ -228,7 +229,94 @@ public Alert apply(String s) {
...
+          .set("first", splits[1])
+          .set("last", splits[2])
+          .build();
+      records.add(record);
+    }
+
+    // Add a null record to get an alert
+    StructuredRecord nullRecord =
+      StructuredRecord.builder(schema)
+        //.set("id", 500L)
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Why is the id not set here ? The key field is `last` and it is set to null.

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java (hunk 3 above)
Comment (sgarg-CS) [reply kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Removed. b6aa798cc20abacecff2b7d570e47d1678017c88

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java (hunk 4 above)
Comment (sgarg-CS) [reply kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: No, it's not required, removed it. b6aa798cc20abacecff2b7d570e47d1678017c88

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java (hunk 5 above)
Comment (sgarg-CS) [reply kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Uncomment this code. `id` is set now. b6aa798cc20abacecff2b7d570e47d1678017c88
Yes, the key field is `last` and it is set to `null`

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java [hunk 6]
```diff
@@ -228,7 +228,86 @@ public Alert apply(String s) {
...
+    sinkProperties.put("brokers", "localhost:" + kafkaPort);
+    sinkProperties.put("referenceName", "kafkaTest");
+    sinkProperties.put("topic", usersTopic);
+    sinkProperties.put("schema", schema.toString());
+    sinkProperties.put("format", "csv");
+    sinkProperties.put("key", "last");
+    sinkProperties.put("async", "FALSE");
+    sinkProperties.put("compressionType", "none");
+
+    Map<String, String> alertProperties = new HashMap<>();
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Remove unused code like  - alertProperties , alertsTopic

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java [hunk 7]
```diff
@@ -228,7 +228,86 @@ public Alert apply(String s) {
...
+
+    DataSetManager<Table> sourceTable = getDataset(inputName);
+    MockSource.writeInput(sourceTable, records);
+
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    try {
+      workflowManager.start();
+      workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 1, TimeUnit.MINUTES);
+    } catch (Exception e) {
+      Assert.assertTrue(workflowManager.getHistory(ProgramRunStatus.FAILED).size() == 1);
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java:304]: Also add a success case that uses `key` field.

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java [hunk 8]
```diff
@@ -228,7 +228,86 @@ public Alert apply(String s) {
...
+      StructuredRecord record =
+        StructuredRecord.builder(schema)
+          .set("id", Long.parseLong(splits[0]))
+          .set("first", splits[1])
+          .set("last", splits[2])
+          .build();
+      records.add(record);
+    }
+
+    // Add a null record to get an alert
```
Comment (greeshmaswaminathan) [inline kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Fix the comment to remove alert

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java (hunk 6 above)
Comment (sgarg-CS) [reply kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Removed. 6d4060a8c4b43dd3d21fa89cbb9672adaf59841a

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java (hunk 8 above)
Comment (sgarg-CS) [reply kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java]: Updated the comment. 6d4060a8c4b43dd3d21fa89cbb9672adaf59841a

File: kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java (hunk 7 above)
Comment (sgarg-CS) [reply kafka-plugins-client/src/test/java/io/cdap/plugin/kafka/KafkaSinkAndAlertsPublisherTest.java:304]: The success case already exists in the previous unit test. `KafkaSinkAndAlertsPublisherTest.testKafkaSinkAndAlertsPublisher()`

**Review by greeshmaswaminathan — APPROVED** [review]

**Comment by sgarg-CS** [comment]
Commits have been squashed. Thank you!
